framework/src/meshgenerators/CircularCorrectionGenerator.C

```diff
@@ -0,0 +1,417 @@
+
+  params.addParam<bool>("move_end_nodes_in_span_direction",
+                        false,
+                        "Whether to move the end nodes in the span direction of the arc or not.");
+
+  params.addRequiredParam<std::vector<BoundaryName>>("input_mesh_circular_boundaries",
+                                                     "Names of the circular boundaries.");
+
+  params.addRangeCheckedParam<std::vector<Real>>(
+      "transition_layer_ratios",
+      "transition_layer_ratios>0&transition_layer_ratios<=1.0",
+      "Ratios to radii as the transition layers on both sides of the original radii.");
+
+  params.addRangeCheckedParam<std::vector<Real>>(
+      "custom_circular_tolerance",
+      "custom_circular_tolerance>0",
+      "Custom tolerances used to verify whether the provided boundaries are circular (default is "
+      "1.0e-6).");
+
+  params.addClassDescription("This CircularCorrectionGenerator object is designed to correct full "
+                             "or partial circular boundaries in a 2D mesh to preserve areas.");
+
+  return params;
+}
+
+CircularCorrectionGenerator::CircularCorrectionGenerator(const InputParameters & parameters)
+  : MeshGenerator(parameters),
+    _input_name(getParam<MeshGeneratorName>("input")),
+    _input_mesh_circular_boundaries(
+        getParam<std::vector<BoundaryName>>("input_mesh_circular_boundaries")),
+    _transition_layer_ratios(isParamValid("transition_layer_ratios")
+                                 ? getParam<std::vector<Real>>("transition_layer_ratios")
+                                 : std::vector<Real>(_input_mesh_circular_boundaries.size(), 0.01)),
+    _custom_circular_tolerance(
+        isParamValid("custom_circular_tolerance")
+            ? (getParam<std::vector<Real>>("custom_circular_tolerance").size() == 1
+                   ? std::vector<Real>(
+                         _input_mesh_circular_boundaries.size(),
+                         getParam<std::vector<Real>>("custom_circular_tolerance").front())
+                   : getParam<std::vector<Real>>("custom_circular_tolerance"))
+            : std::vector<Real>(_input_mesh_circular_boundaries.size(), 1.0e-6)),
+    _move_end_nodes_in_span_direction(getParam<bool>("move_end_nodes_in_span_direction")),
+    _input(getMeshByName(_input_name))
+{
+  if (_transition_layer_ratios.size() != _input_mesh_circular_boundaries.size())
+    paramError("transition_layer_ratios",
+               "this parameter must have the same length as 'input_mesh_circular_boundaries'.");
+  if (_custom_circular_tolerance.size() != _input_mesh_circular_boundaries.size())
+    paramError("custom_circular_tolerance",
+               "if provided, this parameter must have either a unity length or the same length as "
+               "'input_mesh_circular_boundaries'.");
+}
+
+std::unique_ptr<MeshBase>
+CircularCorrectionGenerator::generate()
+{
+  auto input_mesh = dynamic_cast<ReplicatedMesh *>(_input.get());
```

Comment:
This returns `nullptr` if someone is using a `DistributedMesh`.  Test and `mooseError()` for that, rather than segfault?